Repository Cleanup Endpoint

original-brownbear opened the pull request:
* Snapshot cleanup functionality via transport/REST endpoint.
* Added all the infrastructure for this with the HLRC and node client
* Made use of it in tests and resolved relevant TODO
* Added new `Custom` CS element that tracks the cleanup logic. Kept it similar to the delete and in progress classes and gave it some (for now) redundant way of handling multiple cleanups but only allow one
* Use the exact same mechanism used by deletes to have the combination of CS entry and increment in repository state ID provide some concurrency safety (the initial approach of just an entry in the CS was not enough, we must increment the repository state ID to be safe against concurrent modifications, otherwise we run the risk of "cleaning up" blobs that just got created without noticing)
* Isolated the logic to the transport action class as much as I could. It's not ideal, but we don't need to keep any state and do the same for other repository operations (like getting the detailed snapshot shard status)

original-brownbear committed:
Repo cleanup endpoint start

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
50%

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
just ack for now

original-brownbear committed:
just ack for now

original-brownbear committed:
just ack for now

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Fix functionality

original-brownbear committed:
add rest action

original-brownbear committed:
fix compilation

original-brownbear committed:
fix compilation

original-brownbear committed:
nicer formatting

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
merge fixes

original-brownbear committed:
add some documentation

original-brownbear committed:
nicer cleanup logic

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
result response

original-brownbear committed:
compiles

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
fix hdfs repo tests

original-brownbear committed:
add cleanup rest test

original-brownbear committed:
fix rest test

original-brownbear committed:
fix rest test

original-brownbear committed:
shorter logic

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
HLRC integration

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
fix empty repo handling

original-brownbear committed:
Fix HLRC integration

original-brownbear committed:
cleaner

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
ensure no concurrent modifications

original-brownbear committed:
better docs

original-brownbear committed:
fix formatting

original-brownbear committed:
fix formatting

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

elasticmachine commented:
Pinging @elastic/es-distributed

original-brownbear committed:
add javadoc

original-brownbear commented on test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java:
```diff
@@ -226,26 +227,45 @@ public void testCleanup() throws Exception {
                 final BlobStore blobStore = repo.blobStore();
-                blobStore.blobContainer(BlobPath.cleanPath().add("indices").add("foo"))
-                    .writeBlob("bar", new ByteArrayInputStream(new byte[0]), 0, false);
+                blobStore.blobContainer(repo.basePath().add("indices").add("foo"))
```
Moving from clean path to base path actually fixes a bug in this test that became visible through the cleanup result assertion.
The stale index path was created outside the repo base path ... which obviously passed the no-corruption checks regardless of whether or not the cleanup actually worked.

original-brownbear committed:
remove todo that was addressed

original-brownbear committed:
simplify

original-brownbear committed:
revert needless chagne

original-brownbear committed:
remove resolved todo

original-brownbear committed:
ensure cleanup only runs on 8+

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
skip cleanup check in bwc

original-brownbear committed:
skip cleanup check in bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
deal with null spots

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsInProgress.java:
```diff
@@ -140,6 +137,9 @@ public Version getMinimalSupportedVersion() {
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startArray(TYPE);
         for (Entry entry : entries) {
+            if (entry.snapshot.getSnapshotId() == null) {
```
I didn't see much point rendering the cleanup entry here (and in `toString`) since it's not actually a delete.

original-brownbear commented on server/src/main/java/org/elasticsearch/snapshots/Snapshot.java:
```diff
@@ -38,9 +40,9 @@
      * Constructs a snapshot.
      */
-    public Snapshot(final String repository, final SnapshotId snapshotId) {
+    public Snapshot(String repository, @Nullable SnapshotId snapshotId) {
```
This is kinda dirty admittedly, but it seems like it's the shortest route to piggy back onto `SnapshotDeletionsInProgress.Entry`. Unfortunately that class doesn't have a reference to the `repository` otherwise so if I encode this as anything but a `null` `SnapshotId` here I'd have to add the repository to the entry and have it here redundantly.

original-brownbear commented on test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java:
```diff
@@ -227,30 +228,49 @@ public void testCleanup() throws Exception {
-                blobStore.blobContainer(BlobPath.cleanPath().add("indices").add("foo"))
-                    .writeBlob("bar", new ByteArrayInputStream(new byte[0]), 0, false);
+                blobStore.blobContainer(repo.basePath().add("indices").add("foo"))
+                    .writeBlob("bar", new ByteArrayInputStream(new byte[3]), 3, false);
```
This whole logic of checking the count and size is kinda meh being hard coded this way I must admit. A better version of this is incoming in the repo cleanup cli PR and I'd just make use of that to clean things up here too once it's merged if that's ok.

andrershov replied:
This PR is already merged, so I think it's time to re-use methods from BlobStoreTestUtils

original-brownbear replied:
Right ... how time flies :) cleaning this up then :)

original-brownbear replied:
Hmm turns out those won't work without refactoring after all (they don't properly discriminate between folders and blobs so it's more code to use them than not). I'd just clean this up in a follow up (together with a ton of duplication that we have acquired in these tests anyway) since this PR is already 1k+ lines?

original-brownbear commented on test/framework/src/main/java/org/elasticsearch/repositories/AbstractThirdPartyRepositoryTestCase.java:
```diff
@@ -260,10 +280,10 @@ protected boolean assertCorruptionVisible(BlobStoreRepository repo, Executor exe
-                        && blobStore.blobContainer(BlobPath.cleanPath().add("indices").add("foo")).blobExists("bar")
-                        && blobStore.blobContainer(BlobPath.cleanPath()).blobExists("meta-foo.dat")
-                        && blobStore.blobContainer(BlobPath.cleanPath()).blobExists("snap-foo.dat")
+                    blobStore.blobContainer(repo.basePath().add("indices")).children().containsKey("foo")
```
This was an actual bug in the tests! Using clean path here led to the tests passing but not actually testing the creation of a stale index folder (it was created outside the repo path in that case).

original-brownbear commented on server/src/main/java/org/elasticsearch/common/blobstore/BlobContainer.java:
```diff
@@ -111,9 +112,12 @@
 
     /**
      * Deletes this container and all its contents from the repository.
+     * @param resultConsumer Long consumer that implementations can invoke with the byte size of each deleted blob when deleting it.
```
I'm not super happy about having to do it in this fuzzy way here, but I don't see a way of doing it any other way on HDFS without a massive performance hit (HDFS has a recursive delete RPC call that we run one-off but that doesn't return any statistics about what was deleted itself). IMO, this is fine though, for users running this it should mostly be relevant whether or not anything was cleaned up so they can tweak the frequency of running this endpoint accordingly (if they don't experience any failures during snapshotting or deleting snapshots they should never need this endpoint and always getting `0` returns from it will transport that message just fine I hope).

original-brownbear committed:
revert needless changes

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
bck

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleaup-repo-ep-v2

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleaup-repo-ep-v2

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleaup-repo-ep-v2

original-brownbear committed:
resolve conflict

original-brownbear committed:
shorter

original-brownbear committed:
nicer

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleaup-repo-ep-v2

original-brownbear committed:
nicer

original-brownbear committed:
docs + correct handling read-only repo

original-brownbear committed:
More efficient, don't increment index unless there is something to do

original-brownbear committed:
nicer formatting

original-brownbear committed:
add comment

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear commented:
Jenkins run elasticsearch-ci/bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
add some logging

tlrx commented on server/src/main/java/org/elasticsearch/repositories/RepositoryCleanupResult.java:
```diff
@@ -0,0 +1,109 @@
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        return builder.startObject().field("bytes", bytes).field("blobs", blobs).endObject();
```
I'd prefer something like
```json
{
    "deleted_blobs": 123,
    "deleted_blobs_in_bytes": 456789
}
```

original-brownbear replied:
Same here, that looks much nicer, thanks will rename! :)

tlrx commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/TransportCleanupRepositoryAction.java:
```diff
@@ -0,0 +1,251 @@
+     */
+    private void cleanupRepo(String repositoryName, ActionListener<RepositoryCleanupResult> listener) {
+        final Repository repository = repositoriesService.repository(repositoryName);
+        if (repository instanceof BlobStoreRepository == false) {
```
We can check the read only repository upfront here

original-brownbear replied:
You're right ... hmm ... the thing is, we do the check for `readOnly` for all the other operations in `BlobStoreRepository` as well. I can see how it's somewhat strange to go through the whole dance of updating the CS only to then fail because of the read only setting.
On the other hand, we do the check in `BlobStoreRepository` for create and delete and it seems safer to do that before doing any writing than relying on the caller of the cleanup method => maybe it's not worth the duplicate check to optimize the weird corner-case here? (mainly trying to keep the cleanup action code short)

tlrx replied:
Since updating the cluster state can be a heavy task I really think we should fail early. But I'm surprised that we don't early check for read only when creating or deleting a snapshot.

So we could let this at it is in this PR and as a follow we could early check for read only repos in create/delete snapshot and repo clean up. WDYT?

original-brownbear replied:
> So we could let this at it is in this PR and as a follow we could early check for read only repos in create/delete snapshot and repo clean up. WDYT?

Sounds good to me :) Def. makes things a little cleaner.

tlrx commented on server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java:
```diff
@@ -215,12 +216,15 @@ public void deleteBlob(String blobName) {
 
             @Override
-            public void delete() {
+            public void delete(LongConsumer resultConsumer) {
```
I tend to think that this method could return deletion stats, WDYT?

original-brownbear replied:
I tried finding an argument against doing that but couldn't :D Will refactor that :)

tlrx reviewed (commented):
I left a first bunch of comments, mostly minor things. Overall logic looks good but I'd like to take another look.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear commented on client/rest-high-level/src/main/java/org/elasticsearch/client/SnapshotRequestConverters.java:
```diff
@@ -94,6 +95,20 @@ static Request verifyRepository(VerifyRepositoryRequest verifyRepositoryRequest)
+    static Request cleanupRepository(CleanupRepositoryRequest cleanupRepositoryRequest) {
+        String endpoint = new RequestConverters.EndpointBuilder().addPathPartAsIs("_snapshot")
+            .addPathPart(cleanupRepositoryRequest.repository())
+            .addPathPartAsIs("cleanup")
```
Yea, we should IMO. I'll adjust it :)

original-brownbear committed:
CR: resolve some comments

original-brownbear committed:
CR: renamings

original-brownbear committed:
CR: renamings

original-brownbear committed:
CR: fix indent

original-brownbear committed:
CR: order things reasonably

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/TransportCleanupRepositoryAction.java:
```diff
@@ -0,0 +1,251 @@
+        clusterService.submitStateUpdateTask("cleanup repository", new ClusterStateUpdateTask() {
+            @Override
+            public ClusterState execute(ClusterState currentState) {
+                SnapshotDeletionsInProgress deletionsInProgress = currentState.custom(SnapshotDeletionsInProgress.TYPE);
```
Yea, sorry for the weirdness => fixed now :)

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/TransportCleanupRepositoryAction.java:
```diff
@@ -0,0 +1,251 @@
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.GENERIC;
```
Can't do that, we restrict repository operations to generic and snapshot pool only. Since reading the repo state id is the first step in the master action, it wouldn't add anything to have the management pool in the middle of things here.

tlrx replied:
Interesting, the verify repository action uses the management tread pool but quickly fork to the snapshot one. The get repositories one also declares the management thread pool as executor and also reads the repository metadata. Maybe we should make this more uniform.

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/TransportCleanupRepositoryAction.java:
```diff
@@ -0,0 +1,251 @@
+            }
+
+            private void after(@Nullable Exception failure, @Nullable RepositoryCleanupResult result) {
+                if (failure == null) {
```
We do right at the end when the state was processed. This is just a middling step.

tlrx replied:
Right, sorry for the confusion.

original-brownbear committed:
CR: nicer snapshot task name

original-brownbear committed:
CR: nicer naming cleanup CS update tasks

original-brownbear committed:
simplify away some casting

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/TransportCleanupRepositoryAction.java:
```diff
@@ -0,0 +1,251 @@
+                threadPool.executor(ThreadPool.Names.SNAPSHOT).execute(new ActionRunnable<>(listener) {
+                    @Override
+                    protected void doRun() {
+                        Repository repository = repositoriesService.repository(repositoryName);
```
You raise a good point here :)

The cleanup operation is a little less elegant than the snapshot create operation when it comes to handling the repository state id. We read the state id up-front and then create the entry in the CS. So if we have some concurrency here (easily possible) where a snapshot just finishes creating and we still read the old index gen, then we're in the unfortunate situation where the cleanup will fail even though it was set up in the CS and even if there was no master fail-over. 
Just reading again here won't do us much good (compared to what already happens), because if we now see a changed repository state id here we would still have to submit another state update task to align the repo state IDs.

That said: We have the same issue with snapshot deletion as well => I'm wondering if we should just live with it here to keep things simple and then in a follow-up clean up both by adding some notion of `INIT` state to them like we do for snapshot create (so we "lock" the repo before reading the state id)?

original-brownbear replied:
Actually, on second thought I think this is fine. No one ever complained about the equivalent and incredibly unlikely situation for snapshot delete and create either ... We can add stages to the cleanup once we get to deeper cleanup on a shard level imo.

tlrx replied:
I looked at various scenarios and I agree with your reasoning. I also think that reading again from cluster state won't be better after all. 

I think this is fine. On the question of adding some INIT stage to cluster state object markers I need to think about it again.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
CR: rename response fields

original-brownbear commented on server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/cleanup/CleanupRepositoryResponse.java:
```diff
@@ -0,0 +1,76 @@
+        result = new RepositoryCleanupResult(in);
+    }
+
+    public RepositoryCleanupResult result() {
```
Tbh. I don't like "Stats" here:
 We use the "Stats" name in a bunch of spots where we actually get a snapshot of some stats (for caches, snapshots etc.) that are idempotent with respect to the request made (otherwise obviously mostly variable in time). Here that's not really true and it's more of a result isn't it?

original-brownbear commented on server/src/main/java/org/elasticsearch/common/blobstore/fs/FsBlobContainer.java:
```diff
@@ -123,8 +124,22 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOEx
+            @Override
+            public FileVisitResult postVisitDirectory(Path dir, IOException impossible) throws IOException {
+                assert impossible == null;
+                Files.delete(dir);
```
I don't know, should we really? It seems like an exception would always be either a system issue or concurrent operations on the repo. Shouldn't we just stop doing things asap in this case?

tlrx replied:
Right, not sure why I made this comment 3 days ago. Let's barf ealy.

original-brownbear committed:
CR: return a delete result from container deletes

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear committed:
nicer apis

original-brownbear commented on plugins/repository-gcs/src/main/java/org/elasticsearch/repositories/gcs/GoogleCloudStorageBlobStore.java:
```diff
@@ -300,12 +301,15 @@ void deleteBlob(String blobName) throws IOException {
-                page.getValues().forEach(b -> blobsToDelete.add(b.getName()));
+                page.getValues().forEach(b -> {
+                    blobsToDelete.add(b.getName());
+                    resultConsumer.accept(b.getSize());
```
I think now hat we return the delete stats instead of updating some mutable object this is fine. I can see an argument on GCS, that you technically could separate the case of deleting and already missing blob out (not on S3 though) to get more exact counts but I figured that might not be worth the complexity (we could do that in a follow up though, there's some simplifications API wise here that I had planned that would make it easier to do that, but here it seemed like needless added noise)?

tlrx replied:
Good for me, thanks

original-brownbear commented on plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobContainer.java:
```diff
@@ -134,8 +134,11 @@ public void delete() throws IOException {
-                    list.getObjectSummaries().stream().map(S3ObjectSummary::getKey).collect(Collectors.toList());
+                final List<String> blobsToDelete = new ArrayList<>();
+                    list.getObjectSummaries().forEach(s3ObjectSummary -> {
+                        resultConsumer.accept(s3ObjectSummary.getSize());
```
Here it doesn't matter now that we're returning the delete stats, we're throwing if any delete fails anyway so we can keep the simpler code that counts stuff up beforehand imo.

original-brownbear commented:
@tlrx thanks so much for going through this one! I addressed all comments and adjusted the delete API to return an object (sorry for adding another class here, but I figured it was cleaner and/or easier to read in the long run instead of passing around the mutable objects we use for serialisation here).
=> this should be good for another pass I hope :)

original-brownbear commented:
Jenkins run elasticsearch-ci/packaging-sample

ywelsch commented on docs/reference/modules/snapshots.asciidoc:
```diff
@@ -329,6 +329,38 @@ POST /_snapshot/my_unverified_backup/_verify
 
+[float]
+===== Repository Cleanup
+Repositories can over time accumulate data that is not referenced by any existing snapshot. This is a result of the data safety guarantees
```
Mention that the cleanup functionality is also automatically run on snapshot deletion?

original-brownbear replied:
Hmm I'm not sure. It would make sense now, but I was gonna add the logic to clean up all the individual shard folders in a follow-up which we can't run on delete (it's just super slow), not sure if it's worth explaining the difference here?

ywelsch replied:
Most folks shouldn't bother about this API, given that all relevant clean-up will happen on snapshot deletion. I do not want to raise the impression that you will *have* to run this API to sensibly operate  an ES cluster.

original-brownbear replied:
Fair point, I added a note on that now in 
25fb827

andrershov commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -457,12 +503,16 @@ private void cleanupStaleRootFiles(Set<String> rootBlobNames, RepositoryData rep
         ).collect(Collectors.toList());
+    }
+
+    private List<String> cleanupStaleRootFiles(List<String> blobsToDelete) {
```
@original-brownbear do you think we need to add the same logic to the cleanup tool?

original-brownbear replied:
We could add it, I'd expect the effects to be somewhat minor though (unless there is an unknown bug that leads to thousands of leaked files here that add listing costs or a huge cluster state is snapshotted repeatedly) for most repos. But come to think of it ... we had some buggy ones that would never finalise correctly but go through all the other snapshot steps.  I guess I'd check if this is an issue first, then decide if it's worth the added risk?

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

andrershov reviewed (commented):
@original-brownbear Looks good!

I've left a couple of minor comments.

original-brownbear committed:
CR: add note on deletes cleaning up repo as well

original-brownbear commented on docs/reference/modules/snapshots.asciidoc:
```diff
@@ -329,6 +329,38 @@ POST /_snapshot/my_unverified_backup/_verify
+[source,js]
+--------------------------------------------------
+{
+  "results": {
```
I think "results" is more natural language wise here in English (I guess because we're listing multiple non-binary values but I tbh. I don't have a good explanation as to why that is :D).

tlrx replied:
It is the result of the clean up operation, so I think we could just drop the `results`/`result` field, maybe also the RepositoryCleanupResult object, and report 
```
{
    "deleted_bytes": 20,
    "deleted_blobs": 5
}
```

original-brownbear replied:
I mainly did this to stay a little consistent in style with other snapshot APIs. We never really return a flat object, we always wrap under a descriptive key don't we? (e.g. getting a list of snapshots in a repo, creating a snapshot, ...)

Aesthetically I always kind of like an object like that but I don't have a good argument for it otherwise :) I suppose it makes it a little easier to extend if we ever want to in the future, but  that isn't so relevant for this endpoint I guess.

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

tlrx commented on server/src/main/java/org/elasticsearch/common/blobstore/DeleteResult.java:
```diff
@@ -0,0 +1,52 @@
+
+    public static final DeleteResult ZERO = new DeleteResult(0, 0);
+
+    private final long blobsDeleted;
```
nit: this could be just `blobs` & `bytes`

original-brownbear replied:
Hmm, I'd rather not, then I have to use `this` in `add` :D

tlrx commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestCleanupRepositoryAction.java:
```diff
@@ -0,0 +1,56 @@
+ * Cleans up a repository
+ */
+public class RestCleanupRepositoryAction extends BaseRestHandler {
+    public RestCleanupRepositoryAction(Settings settings, RestController controller) {
```
nit: missing space

tlrx reviewed (commented):
Thanks for the changes, I liked it better without the LongConsumer. Can we decide on the response output? That's the last item that prevents me to LGTM ;)

Also, did you consider to have a dry-run option?

original-brownbear commented:
Thanks @tlrx 

> Thanks for the changes, I liked it better without the LongConsumer. 

Np, so do I :D

> Can we decide on the response output? That's the last item that prevents me to LGTM ;)

I like the one level of nesting better but have no strong arguments for it other than maybe consistency with other APIs and extensibility.

> Also, did you consider to have a dry-run option?

Yea but in the end I'm not sure I see the point. If you think about it, all you can present to the user is a bunch of uuids for indices and stale root blobs). What can they really get out of that?
On the other hand, the response could be problematically huge in some cases and if we add more logic here to do some deep cleaning of shards it gets questionable how we would even present the results of that (IMO) -> not sure it's worth it unless someone sees a good use case?

original-brownbear committed:
add missing empty line

original-brownbear commented:
Jenkins run elasticsearch-ci/bwc

original-brownbear commented:
Jenkins run elasticsearch-ci/bwc

original-brownbear commented:
Jenkins run elasticsearch-ci/bwc

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

andrershov reviewed (approved):
LGTM

original-brownbear committed:
Merge remote-tracking branch 'elastic/master' into cleanup-repo-ep

original-brownbear commented:
Jenkins run elasticsearch-ci/packaging-sample

original-brownbear commented:
@tlrx ping :) (no rush I know you just returned)

> Can we decide on the response output? That's the last item that prevents me to LGTM ;)

I still like the current format :) ... mainly for its easier extensibility. If that's the last thing blocking this maybe we can go with the current version and merge this? :)

tlrx reviewed (approved):
LGTM

Concerning the response output I expressed my opinion but if both you and @andrershov prefer the proposed format then I'm fine :)

Thanks for the response on the dry-run option, I wanted to know if it was considered at some point. Since this is something that can be added later, we'll see if someone requests it.

original-brownbear commented:
Thanks so much @tlrx and @andrershov for reviewing this big one!
